import pandas as pd
import pytest
from heliolib.utils.time_zone_info_handler import TimeZoneInfoHandler


def prepare_timezone_naive_dataframe():
    """Returns a DataFrame with a timezone-naive DatetimeIndex."""
    return pd.DataFrame(
        {"data": range(3)}, index=pd.date_range("2023-01-01", periods=3)
    )


def prepare_utc_aware_dataframe():
    """Returns a DataFrame with a timezone-aware DatetimeIndex."""
    return pd.DataFrame(
        {"data": range(3)}, index=pd.date_range("2023-01-01", periods=3, tz="UTC")
    )


def test_make_naive_datetime_specific_timezone_aware_validates_with_time_zone_naive_index():
    time_zone_naive_dataframe = prepare_timezone_naive_dataframe()
    timezone_string = "Europe/London"
    timezone_handler = TimeZoneInfoHandler()
    timezone_handler.make_naive_datetime_specific_timezone_aware(
        timezone_str=timezone_string, dataframe=time_zone_naive_dataframe
    )
    assert str(time_zone_naive_dataframe.index.tzinfo) == timezone_string


def test_make_naive_datetime_specific_timezone_aware_fails_with_utc_aware_index():
    utc_aware_dataframe = prepare_utc_aware_dataframe()
    timezone_string = "Europe/London"
    timezone_handler = TimeZoneInfoHandler()
    with pytest.raises(Exception) as exc_info:
        timezone_handler.make_naive_datetime_specific_timezone_aware(
            timezone_str=timezone_string, dataframe=utc_aware_dataframe
        )
    assert (
        str(exc_info.value)
        == "An error occurred while making the datetime index Europe/London aware: Datetime index is already timezone aware."
    )
